### PR TITLE
[MLv2] Don't allow adding an expression with the same name as a column

### LINKED ISSUE #30401: [MLv2] [Bug] Field ref names can become ambiguous
Notice that the `order-by` `:field`s are both using `"ID"` whereas the last one should be using `"ID_2"`. 

```clojure
(let [query (-> (lib/query-for-table-name meta/metadata-provider "CATEGORIES")
                (lib/expression "ID" (lib/field "CATEGORIES" "NAME")) ;; can use "ID" but this shows different types below
                (lib/append-stage))
      cols (lib/orderable-columns query)]
  (-> query
      (lib/order-by (lib/ref (first cols)))
      (lib/order-by (lib/ref (last cols)))))
      
{:database 19001,
 :stages [{:expressions {"ID" [:field
                               {:base-type :type/Text,
                                :effective-type :type/Text,
                                :lib/uuid "931484f4-fab8-4c17-8a37-7f1b478c05b2"}
                               19101]},
           :source-table 19010,
           :lib/options {:lib/uuid "b99ef6ce-6caa-4267-bbbf-a21240fccff5"},
           :lib/type :mbql.stage/mbql}
          {:order-by [[:asc
                       {:lib/uuid "a800c72a-8513-490a-94ee-6d19e0b7f1e6"}
                       [:field
                        {:base-type :type/BigInteger,
                         :effective-type :type/BigInteger,
                         :lib/uuid "9a40250d-abf8-4598-9dcb-0d27e79f1c77"}
                        "ID"]]
                      [:asc
                       {:lib/uuid "ba2971b4-0ad4-4b83-8ec3-8283f262cc56"}
                       [:field
                        {:base-type :type/Text,
                         :effective-type :type/Text,
                         :lib/uuid "8fb8b4bb-fb92-47b0-a180-1fb0c83d9161"}
                        "ID"]]],
           :lib/type :mbql.stage/mbql}],
 :lib/type :mbql/query}
 ```

### PR DESCRIPTION
**Pull request by bshepherdson**
This was causing ambiguous references, since `order-by` clauses
reference columns from previous stages by name.

Fixes #30401.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30809)

### CODE REVIEW

File: src/metabase/lib/expression.cljc [hunk 1]
```diff
@@ -184,12 +185,22 @@
   [query stage-number [_coalesce _opts expr _null-expr]]
   (lib.metadata.calculation/column-name query stage-number expr))
 
+(defn- unique-name? [query stage-number expression-name]
+  (let [stage (lib.util/query-stage query stage-number)
+        cols  (lib.metadata.calculation/visible-columns query stage-number stage)
+        names (set (for [{col :name} cols]
+                     (u/lower-case-en col)))]
+    (not (names (u/lower-case-en expression-name)))))
```
Comment (metamben) [inline src/metabase/lib/expression.cljc:193]:
```suggestion
(defn- unique-name? [query stage-number expression-name]
  (let [stage (lib.util/query-stage query stage-number)
        cols  (lib.metadata.calculation/visible-columns query stage-number stage)
        expr-name (u/lower-case-en expression-name)]
    (not-any? #(= (u/lower-case-en (:name %)) expr-name) cols)))
```

File: src/metabase/lib/expression.cljc [hunk 2]
```diff
@@ -184,12 +185,22 @@
...
+
 (mu/defn expression :- ::lib.schema/query
   "Adds an expression to query."
   ([query expression-name an-expression-clause]
    (expression query -1 expression-name an-expression-clause))
   ([query stage-number expression-name an-expression-clause]
    (let [stage-number (or stage-number -1)]
+     (when-not (unique-name? query stage-number expression-name)
+       (throw (ex-info "Expression name conflicts with a column in the same query stage"
```
Comment (metamben) [inline src/metabase/lib/expression.cljc:202]: It might make sense to include the conflicting name in the message.

**Review by metamben — APPROVED** [review]
LGTM

File: src/metabase/lib/expression.cljc [hunk 3]
```diff
@@ -184,12 +185,22 @@
...
+                     (u/lower-case-en col)))]
+    (not (names (u/lower-case-en expression-name)))))
+
 (mu/defn expression :- ::lib.schema/query
   "Adds an expression to query."
   ([query expression-name an-expression-clause]
    (expression query -1 expression-name an-expression-clause))
   ([query stage-number expression-name an-expression-clause]
    (let [stage-number (or stage-number -1)]
+     (when-not (unique-name? query stage-number expression-name)
```
Comment (metamben) [inline src/metabase/lib/expression.cljc:201]: Might test for clash instead and save two negations.

File: src/metabase/lib/expression.cljc (hunk 2 above)
Comment (bshepherdson) [reply src/metabase/lib/expression.cljc:203]: Done.

**Pull request merged**